### PR TITLE
doc: Add api curl examples

### PR DESCRIPTION
**Pull request by benolayinka**
#### Summary

Add some more `curl` http examples

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Run Locally: Verified that the docs build using `make server`
- [ ] New Features Marked: Documentation for new features is marked using the `new-in-version` shortcode, according to the guidelines in `README.md`.
- [x] Style Guidelines: Documentation obeys style guidelines in `README.md`.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/reference/api/_index.md
Comment (htdvisser) [inline doc/content/reference/api/_index.md:33]: Do we have any guidelines for splitting commands over multiple lines?

File: doc/content/reference/api/_index.md
Comment (htdvisser) [inline doc/content/reference/api/_index.md:22]: I think we typically say "multi-tenant deployments"

File: doc/content/reference/api/_index.md
Comment (htdvisser) [inline doc/content/reference/api/_index.md:30]: It's not a list, it's a stream

File: doc/content/reference/api/_index.md
Comment (htdvisser) [inline doc/content/reference/api/_index.md:35]: The content type of the request is definitely not `text/plain`. We're sending JSON.

File: doc/content/reference/api/_index.md
Comment (htdvisser) [inline doc/content/reference/api/_index.md:50]: I don't think we shorten IS, JS, NS, AS in documentation.

File: doc/content/reference/api/_index.md
Comment (htdvisser) [inline doc/content/reference/api/_index.md:57]: Not `text/plain`. We're sending JSON.

File: doc/content/reference/api/_index.md
Comment (htdvisser) [reply doc/content/reference/api/_index.md:35]: And we're expecting JSON back, so we should probably also have `Accept` for JSON.

**Review by htdvisser — CHANGES_REQUESTED** [review]
I really don't think the index page of our API reference should document how you can create end devices with cURL, which we don't want users to do in the first place

**Comment by benolayinka** [comment]
> I really don't think the index page of our API reference should document how you can create end devices with cURL, which we don't want users to do in the first place

I'm split here. It's a robust example, and even if we don't recommend that people do it, it provides a lot of insight about how to interact with the HTTP API. @johanstokking  ?

File: doc/content/reference/api/_index.md
Comment (johanstokking) [reply doc/content/reference/api/_index.md:33]: Also missing `$` here that we use elsewhere

**Comment by johanstokking** [comment]
Yes, I do think it makes sense to document device provisioning the HTTP way, i.e. with cURL examples.

**Comment by htdvisser** [comment]
I'm not against documenting it. I just think it shouldn't be on the **index page** of our API reference

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:36]: Why no `Accept` header in the two requests above?

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:52]: 1. Refer to the API reference to see which fields belong on which server.
2. We don't write "Identity Server" etc. as `inline code`

**Review by htdvisser — CHANGES_REQUESTED** [review]

File: doc/content/getting-started/api/_index.md
Comment (benolayinka) [reply doc/content/getting-started/api/_index.md:52]: Just took them from the console

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:50]: We need to tell people which fields go where. This is already described in the API reference, so please link to `reference/api/end_device`.

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:56]: Let's be consistent and also send the `Accept` header here and below.

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:52]: This is also where you register a friendly name, description, attributes, locations, pictures etc. Again: the exact fields are described in the API reference.

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:156]: I'm pretty sure these are optional (correct me if I'm wrong, @rvolosatovs), so you can omit them (and the field paths below) to simplify this call a bit.

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:81]: This doesn't really describe what's happening below, where we're not actually setting the `mac_setting`.

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:96]: Not sure if we should confuse people with Class C in a "Getting Started". Setting it to `true` when not actually true, will probably cause confusion.

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [reply doc/content/getting-started/api/_index.md:21]: Probably nicer to use newlines for each argument, for readability.

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:37]: Here you use quotes around the URL, not above. Please pick one. If you do, then you don't need to escape `?`

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:50]:
```suggestion
If you want to register a device via the HTTP API, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. 
```

It's highly unlikely that people will use cURL _manually_; these are mostly examples for developers to integrate in their software for automated onboarding. That is a very valid reason to use the HTTP API directly.

**Review by johanstokking — CHANGES_REQUESTED** [review]
Also agreed with @htdvisser comments

File: doc/content/getting-started/api/_index.md
Comment (rvolosatovs) [reply doc/content/getting-started/api/_index.md:156]: Indeed, no need to set zero values (so no `null` and no `""`)

File: doc/content/getting-started/api/_index.md
Comment (rvolosatovs) [inline doc/content/getting-started/api/_index.md:160]: perhaps align this with the actual key length?

File: doc/content/getting-started/api/_index.md
Comment (rvolosatovs) [inline doc/content/getting-started/api/_index.md:150]: For `key` you use `X`s, here you use `0`s, I'd prefer a consistent format for context-specific data, perhaps just use `X` to fool-proof this, such that people don't start copy-pasting commands and creating a bunch of 0000000000000000 DevEUI and JoinEUI devices?

File: doc/content/getting-started/api/_index.md
Comment (rvolosatovs) [reply doc/content/getting-started/api/_index.md:96]: Yes, no need to set this field at all, just assume class A and use the minimum amount of data required

File: doc/content/getting-started/api/_index.md
Comment (rvolosatovs) [inline doc/content/getting-started/api/_index.md:101]: I would prefer only the required fields here

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

File: doc/content/getting-started/api/_index.md
Comment (benolayinka) [reply doc/content/getting-started/api/_index.md:4]: will take a look at the whole section in another pr https://github.com/TheThingsIndustries/lorawan-stack-docs/issues/73

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:45]: Does this work; order of flags, url, more flags?

Comment (johanstokking) [reply doc/content/getting-started/api/_index.md:45]: My help says;

```
Usage: curl [options...] <url>
```

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:57]: One pair of `()` too many

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:59]: Idem

**Review by htdvisser — COMMENTED** [review]
Not much to add.

Maybe also good to document some best practices:

- Setting a `User-Agent` header containing the integration name and version. That way, a network operator can target specific integrations and versions in their firewall rules in case some integration misbehaves, instead of blocking all versions of `curl` or your integration completely.
- Respecting rate limits, meaning to support the `X-RateLimit-*` headers (check with @neoaggelos) and act accordingly. That way a network operator won't have to return `429 Too Many Requests` errors, or even block your integration.
- Supporting the `X-Warning` header. That way you can notice any issues with your integration that may result in errors in the future.

File: doc/content/getting-started/api/_index.md
Comment (benolayinka) [reply doc/content/getting-started/api/_index.md:45]: it does work but it's semantically wrong, i'll rewrite

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:61]:
```suggestion
If you want to do something like registering a device directly via the API, you need to make calls to the Identity Server, Join Server, Network Server and Application Server. See the [API Reference]({{< ref "reference/api" >}}) for detailed information about which messages go to which endpoints. 
```

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:197]:
```suggestion
Set the `User-Agent` HTTP header containing your integration name and version. That way, a network operator can help finding out potential issues using the logs.
```

`User-Agent` should never be used for security

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:199]:
```suggestion
### Respect `X-Ratelimit-*` Response Headers
```

File: doc/content/getting-started/api/_index.md
Comment (johanstokking) [inline doc/content/getting-started/api/_index.md:203]:
```suggestion
### Mind the `X-Warning` Headers
```

**Review by johanstokking — APPROVED** [review]
Approved with minor comments

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:201]: It feels a bit weird to refer to the API documentation of Auth0 for this. 

If we want to provide background, we could point to the internet-draft (https://tools.ietf.org/id/draft-polli-ratelimit-headers-03.html) and explain that we use the `X-` prefix until it becomes an RFC.

File: doc/content/getting-started/api/_index.md
Comment (htdvisser) [inline doc/content/getting-started/api/_index.md:44]: The event stream accept header should be `Accept: text/event-stream`. See also the API reference: https://thethingsstack.io/reference/api/events/

Comment (htdvisser) [reply doc/content/getting-started/api/_index.md:44]: And maybe we should document what a `text/event-stream` response looks like. 

Background: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events

**Review by htdvisser — APPROVED** [review]
LGTM. Before merging, please check if that `User-Agent` header is correct. I'm not sure if it allows space around the `/`.

File: doc/content/getting-started/api/_index.md
Comment (rvolosatovs) [inline doc/content/getting-started/api/_index.md:159]: I guess it would be nice to mention that this is only relevant for OTAA devices?

**Review by rvolosatovs — APPROVED** [review]
I don't think waiting for my review was really necessary here, approval from @htdvisser and @johanstokking should be enough. LGTM anyway

**Pull request merged**
